### PR TITLE
Rich Text Editor | Added HardBreak extension in editor for force line break (<br> tag)

### PR DESCRIPTION
**Pull request by aagash-ni**
# Pull Request

## 🤨 Rationale

Fix for [Bug 2516897](https://dev.azure.com/ni/DevCentral/_workitems/edit/2516897): Switching to a new line by pressing enter adds more spaces than expected.

- Pressing enter in the editor for a new line adds a `Paragraph tag` which has more space than expected between lines.
-  Enabling [HardBreak ](https://tiptap.dev/api/nodes/hard-break) will allow the user to switch to a new line by pressing `Shift/Ctrl/Cmd + Enter` which will make minimal space between lines by adding `<br> tag`  instead of `Paragraph Tag`.

## 👩‍💻 Implementation

- Added [`HardBreak` ](https://tiptap.dev/api/nodes/hard-break) node in Tiptap Editor.
- Enabled [`newline`](https://github.com/markdown-it/markdown-it/blob/2b6cac25823af011ff3bc7628bc9b06e483c5a08/lib/rules_inline/newline.js#L27C8-L28C42) and [`escape`](https://github.com/markdown-it/markdown-it/blob/2b6cac25823af011ff3bc7628bc9b06e483c5a08/lib/rules_inline/escape.js#L1) rules in MarkdownIt for `RichTextMarkdownParser` to parse the HardBreak markdown syntax as per [CommonMark Spec](https://spec.commonmark.org/0.30/#hard-line-breaks) to `<br> tag`.
- Added `hardBreak` node to `RichTextMarkdownSerializer` to serialize the `<br> tag` to respective hard break markdown syntax.
- As we are using `prosemirror-markdown`, it serializes the `<br>` tag to backslash with [line ending](https://spec.commonmark.org/0.30/#hard-line-breaks) syntax as per the [CommonMark Spec example 633](https://spec.commonmark.org/0.30/#example-633).

In Windows `Shift/Ctrl+ Enter` and in Mac `Cmd + Enter` will switch to a new line (line break).

We are not exposing any format buttons to add line breaks (Hard Break), The only way to break the line is by using the Keys mentioned above.

## 🧪 Testing

- Added unit tests and visual tests for the functionality.
- Manually tested and verified the functionality of the supported features in Storybook build.

## ✅ Checklist

- [x] I have updated the project documentation to reflect my changes or determined no changes are needed.

### CODE REVIEW

File: packages/nimble-components/src/rich-text/models/tests/markdown-parser.spec.ts [hunk 1]
```diff
@@ -344,4 +344,52 @@ describe('Markdown parser', () => {
...
+    describe('Markdown string with hard break should have respective br tag when rendered', () => {
+        const focused: string[] = [];
+        const disabled: string[] = [];
+        const markdownStringWithHardBreak: {
+            name: string,
+            tags: string[]
+        }[] = [
+            {
+                name: '**bold**\\\n*Italics*',
+                tags: ['P', 'STRONG', 'BR', 'EM']
```
Comment (vikisekarNI) [inline packages/nimble-components/src/rich-text/models/tests/markdown-parser.spec.ts:357]: Can we replace name here as `value` and give a valid test case title with `name` which is easy to understand?

Comment (vikisekarNI) [reply packages/nimble-components/src/rich-text/models/tests/markdown-parser.spec.ts:357]: For ex: 
1. `should render br tag between bold and italics`
2. `should render br tag between two first level bullet list items`

File: packages/nimble-components/src/rich-text/models/tests/markdown-serializer.spec.ts [hunk 2]
```diff
@@ -258,4 +255,71 @@ describe('Markdown serializer', () => {
...
+                    '1. list\\\n   hard break content\n\n2. list\n\n   1. nested list\\\n      nested hard break content'
+            }
+        ];
+
+        const focused: string[] = [];
+        const disabled: string[] = [];
+        for (const value of supportedNodesMarks) {
+            const specType = getSpecTypeByNamedList(value, focused, disabled);
+            specType(
+                `Should return ${value.name} markdown (${value.markdown}) when its respective node is passed`,
```
Comment (vikisekarNI) [inline packages/nimble-components/src/rich-text/models/tests/markdown-serializer.spec.ts:314]:
```suggestion
                `Should serialize ${value.name} to markdown`,
```

File: packages/nimble-components/src/rich-text/editor/tests/rich-text-editor.spec.ts [hunk 3]
```diff
@@ -335,6 +344,20 @@ describe('RichTextEditor', () => {
             expect(pageObject.getEditorLeafContents()).toEqual(['bold']);
         });
 
+        it('should have br tag name when pressing shift + Enter with bold content', async () => {
+            await pageObject.clickFooterButton(ToolbarButton.bold);
+            await pageObject.setEditorTextContent('bold1');
+            await pageObject.pressShiftEnterKeysInEditor();
+            await pageObject.setEditorTextContent('bold2');
```
Comment (vikisekarNI) [inline packages/nimble-components/src/rich-text/editor/tests/rich-text-editor.spec.ts:351]:
```suggestion
            await pageObject.setEditorTextContent('bold after hard break');
```

File: packages/nimble-components/src/rich-text/editor/tests/rich-text-editor.spec.ts [hunk 4]
```diff
@@ -343,6 +366,20 @@ describe('RichTextEditor', () => {
             expect(pageObject.getEditorLeafContents()).toEqual(['italics']);
         });
 
+        it('should have br tag name when pressing shift + Enter with Italics content', async () => {
+            await pageObject.clickFooterButton(ToolbarButton.italics);
+            await pageObject.setEditorTextContent('italics1');
+            await pageObject.pressShiftEnterKeysInEditor();
+            await pageObject.setEditorTextContent('italics2');
```
Comment (vikisekarNI) [inline packages/nimble-components/src/rich-text/editor/tests/rich-text-editor.spec.ts:373]:
```suggestion
            await pageObject.setEditorTextContent('italics after hard break');
```

Comment (vikisekarNI) [reply packages/nimble-components/src/rich-text/editor/tests/rich-text-editor.spec.ts:373]: nit: Apply similarly for other test cases for a slightly enhanced readability

File: packages/nimble-components/src/rich-text/editor/tests/rich-text-editor.spec.ts [hunk 5]
```diff
@@ -890,6 +1038,39 @@ describe('RichTextEditor', () => {
         }
     });
 
+    describe('Should return markdown with back slash (HardBreak) when markdown with hard break syntax are assigned', () => {
```
Comment (vikisekarNI) [inline packages/nimble-components/src/rich-text/editor/tests/rich-text-editor.spec.ts:1041]: Just curious to understand why you wanted to have this test case now? 

Title suggestion
```suggestion
    describe('`getMarkdown` with hard break backslashes should be same immediately after `setMarkdown`', () => {
```

File: packages/nimble-components/src/rich-text/models/markdown-parser.ts [hunk 6]
```diff
@@ -43,7 +43,9 @@ export class RichTextMarkdownParser {
         const supportedTokenizerRules = zeroTokenizerConfiguration.enable([
             'emphasis',
             'list',
-            'autolink'
+            'escape',
+            'autolink',
```
[the remark this answers is not in the document]
Comment (vikisekarNI) [reply packages/nimble-components/src/rich-text/models/markdown-parser.ts:47]: Never mind. My bad. I missed to see the autolink entry above

File: packages/nimble-components/src/rich-text/models/tests/markdown-parser.spec.ts (hunk 1 above)
Comment (aagash-ni) [reply packages/nimble-components/src/rich-text/models/tests/markdown-parser.spec.ts]: Updated

File: packages/nimble-components/src/rich-text/models/tests/markdown-serializer.spec.ts (hunk 2 above)
Comment (aagash-ni) [reply packages/nimble-components/src/rich-text/models/tests/markdown-serializer.spec.ts]: Updated

File: packages/nimble-components/src/rich-text/editor/tests/rich-text-editor.spec.ts (hunk 4 above)
Comment (aagash-ni) [reply packages/nimble-components/src/rich-text/editor/tests/rich-text-editor.spec.ts]: Updated as per your suggestion.

File: packages/nimble-components/src/rich-text/editor/tests/rich-text-editor.spec.ts (hunk 5 above)
Comment (aagash-ni) [reply packages/nimble-components/src/rich-text/editor/tests/rich-text-editor.spec.ts]: Updated

File: packages/nimble-components/src/rich-text/editor/tests/rich-text-editor.spec.ts (hunk 3 above)
Comment (aagash-ni) [reply packages/nimble-components/src/rich-text/editor/tests/rich-text-editor.spec.ts]: Updated

File: packages/nimble-components/src/rich-text/models/markdown-serializer.ts [hunk 7]
```diff
@@ -47,7 +47,8 @@ export class RichTextMarkdownSerializer {
             orderedList: orderedListNode,
             doc: defaultMarkdownSerializer.nodes.doc!,
             paragraph: defaultMarkdownSerializer.nodes.paragraph!,
-            text: defaultMarkdownSerializer.nodes.text!
+            text: defaultMarkdownSerializer.nodes.text!,
```
[the remark this answers is not in the document]
Comment (aagash-ni) [reply packages/nimble-components/src/rich-text/models/markdown-serializer.ts:50]: Just added `,` to `text` line, So it's showing changes. It is already used in previous implementations.

File: packages/nimble-components/src/rich-text/editor/tests/rich-text-editor.spec.ts [hunk 8]
```diff
@@ -890,6 +1046,39 @@ describe('RichTextEditor', () => {
...
+            { name: '**bold**\\\n*Italics*' },
+            { name: '* list\\\n  hard break content\n\n* list' },
+            {
+                name: '* list\\\n  hard break content\n\n* list\n\n  * nested list\\\n    nested hard break content'
+            },
+            { name: '1. list\\\n   hard break content\n\n2. list' },
+            {
+                name: '1. list\\\n   hard break content\n\n2. list\n\n   1. nested list\\\n      nested hard break content'
+            }
+        ];
```
Comment (vikisekarNI) [inline packages/nimble-components/src/rich-text/editor/tests/rich-text-editor.spec.ts:1060]: Consider having a value here to keep the test case title simple as mentioned in [other comment](https://github.com/ni/nimble/pull/1517#discussion_r1323261302)

File: packages/nimble-components/src/rich-text/models/tests/markdown-parser.spec.ts [hunk 9]
```diff
@@ -344,4 +344,59 @@ describe('Markdown parser', () => {
...
+                name: 'two first level numbered list items and with nested list',
+                value: '1. list\\\nhard break content\n2. list \n   1. nested list\\\nnested hard break content',
+                tags: ['OL', 'LI', 'P', 'BR', 'LI', 'P', 'OL', 'LI', 'P', 'BR']
+            }
+        ];
+
+        for (const value of markdownStringWithHardBreak) {
+            const specType = getSpecTypeByNamedList(value, focused, disabled);
+            specType(
+                `should render br tag between "${value.name}"`,
```
Comment (vikisekarNI) [inline packages/nimble-components/src/rich-text/models/tests/markdown-parser.spec.ts:391]:
```suggestion
                `should render br tag with "${value.name}"`,
```

**Review by vikisekarNI — APPROVED** [review]

File: packages/nimble-components/src/rich-text/models/tests/markdown-parser.spec.ts (hunk 9 above)
Comment (aagash-ni) [reply packages/nimble-components/src/rich-text/models/tests/markdown-parser.spec.ts]: Updated

File: change/@ni-nimble-components-90bda8e4-208e-4e0c-ae48-b3730a529c57.json [hunk 10]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (aagash-ni) [inline change/@ni-nimble-components-90bda8e4-208e-4e0c-ae48-b3730a529c57.json:1]: I have added `@tiptap/extension-hard-break` package for force line break and will have keyboard accessibility `Ctrl/Shift/Cmd + Enter`, So It needs to be updated in the spec file.

Whether we can update the spec file in this same PR or separate PR is required? @jattasNI @rajsite

**Comment by m-akinc** [comment]
As a reminder, please create the PR in draft mode and let a buddy (me) review before publishing and adding owners.

File: packages/nimble-components/src/rich-text/editor/testing/rich-text-editor.pageobject.ts [hunk 11]
```diff
@@ -114,7 +126,16 @@ export class RichTextEditorPageObject {
...
+        const breakTags = lastElement!.parentElement!.querySelectorAll('br');
+        if (breakTags && breakTags.length > 1) {
+            const textNode = document.createTextNode(value);
+            lastElement!.parentElement!.insertBefore(
+                textNode,
+                lastElement as Node
+            );
+        } else {
+            lastElement!.parentElement!.textContent = value;
+        }
```
[the remark this answers is not in the document]
Comment (m-akinc) [reply packages/nimble-components/src/rich-text/editor/testing/rich-text-editor.pageobject.ts:144]: I'm confused by expected behavior of this function. In the old implementation (and the `else` case in the new implementation), setting the `textContent` of the `parentElement` overwrites anything that might have been there, including the `lastElement`, right? Like in the following situation, the new text would replace both existing strings, right?
image
 But if there are `<br>` elements this function appends instead? For simplicity and consistency, maybe this should always append:
```suggestion
        const textNode = document.createTextNode(value);
        lastElement!.parentElement!.appendChild(textNode);
```
It looks like any placeholder elements (e.g. `ProseMirror-trailingBreak`) get removed when you insert after them.

**Review by m-akinc — APPROVED** [review]

File: packages/nimble-components/src/rich-text/editor/tests/rich-text-editor.spec.ts [hunk 12]
```diff
@@ -912,6 +1068,47 @@ describe('RichTextEditor', () => {
         }
     });
 
+    describe('`getMarkdown` with hard break backslashes should be same immediately after `setMarkdown`', () => {
+        const hardBreakMarkdownStrings: { name: string, value: string }[] = [
+            { name: 'bold and italics', value: '**bold**\\\n*Italics*' },
```
[the remark this answers is not in the document]
Comment (aagash-ni) [reply packages/nimble-components/src/rich-text/editor/tests/rich-text-editor.spec.ts:1073]: In `String.raw` `\n` is considered as raw string, So in cases containing new lines, will look like the below image when we add `String.raw`. 

Still, we can add this if it's fine for you. Please let me know your thoughts.

image

File: packages/nimble-components/src/rich-text/editor/tests/rich-text-editor.spec.ts [hunk 13]
```diff
@@ -418,6 +471,25 @@ describe('RichTextEditor', () => {
             ).toBeTrue();
         });
 
+        it('should have br tag name when pressing shift + Enter with nested numbered lists content', async () => {
```
[the remark this answers is not in the document]
Comment (aagash-ni) [reply packages/nimble-components/src/rich-text/editor/tests/rich-text-editor.spec.ts:474]: Removed a few cases where the hard breaks are checked in a mixed nested list( Bulleted list with nested numbered list and vice versa ). I had other cases where I checked the first level of the nested list just to verify the DOM.

**Comment by atmgrifter00** [comment]
You can bypass me for this review.

File: packages/nimble-components/src/rich-text/editor/tests/rich-text-editor.spec.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (m-akinc) [reply packages/nimble-components/src/rich-text/editor/tests/rich-text-editor.spec.ts:1029]: Let's go with that. At least then it's easier to see how whitespace affects alignment, and the backslashes stand out better.

**Review by rajsite — COMMENTED** [review]
Waiting to see results of test changes

File: packages/nimble-components/src/rich-text/editor/tests/rich-text-editor.spec.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (aagash-ni) [reply packages/nimble-components/src/rich-text/editor/tests/rich-text-editor.spec.ts]: Updated

**Comment by aagash-ni** [comment]
> Waiting to see results of test changes

Hi @rajsite , I hope you are waiting for the Mert test change comments to be addressed, please let us know if you have any other concerns. I have addressed all the comments that Mert mentioned.

**Review by jattasNI — APPROVED** [review]

**Review by rajsite — APPROVED** [review]

**Comment by rajsite** [comment]
Can bypass @mollykreis and @atmgrifter00, so ready to merge!

**Pull request merged**